### PR TITLE
Don't allow lane connections between tracks disconnected due to MaxTurnAngle

### LINKED ISSUE #293: Lane connector cannot be used to make U-turns on medium/large roads with monorail
Lane connector cannot be used to make U-turns on medium and large roads with monorail. This applies to stock and workshop assets. The outgoing connector does not appear on the opposite side of the road. I'm not sure whether it is not rendered or just simply does not allow to enable U-turns

edit: details

### LINKED ISSUE #649: V11 - Could not load type 'TrafficManager.Traffic.Data.VehicleState' from assembly 'TrafficManager`
> # Solution
>   
> Unsubscribe / remove the following mods:
> * **CSUR_TMPE_Laneconnector_Fix**
> * **Advanced Junction Rule**
> ----

### Describe the problem
After the V11 update, I entered the game as before, and an error message popped up。
Could not load type 'TrafficManager.Traffic.Data.VehicleState' from assembly 'TrafficManager, Version=1.0.7026.40416, Culture=neutral, PublicKeyToken=null'. [System.TypeLoadException]

Details:
No details

I subscribed to the V11LABS version. No other subscriptions. There is no problem returning 10.20. Find solution: subscribe to V11 and subscribe to 10.20, the problem will only disappear when V11 is enabled

### Steps to reproduce

1. This error is caused by V11 update, no other modules are added
2. 
3. 

### Savegame?

https://steamcommunity.com/sharedfiles/filedetails/?id=1988175689

### Notes or questions?
[TMPE.log](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/4151649/TMPE.log)
[output_log.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/4151651/output_log.txt)

### PR DESCRIPTION
**Pull request by kianzarrin**
TM:PE lane connector was allowing connections between tracks that were disconnected at junctions due to asset-defined MaxTurnAngle setting. This PR fixes that.

fixes #649
Fixes #293 see comments.
### Images:
 
- https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583849018
- https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583911208

### Related:

#243 . see https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583915848

- I Tidied up the logic for drawlane. its more or less the same but it makes a exception to allow roads. 
- I also made an exception to disallow  lanes that cannot be connected to anything. see - https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583849018
- CSUR needs to read TMPE assembly version to check whether it is necessary to patch TMPE:
https://github.com/citiesskylines-csur/CSURToolBox/pull/1 . So I changed assembly version to 11.1.1 . this will create a merge conflict with #680 which I can easily resolve.

### CODE REVIEW

**Comment by aubergine10** [comment]
FYI, when creating a pull request you can create it as a DRAFT - people can still see it, comment, build and test, but it will prevent merging until it is marked ready for review: https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by krzychu124** [comment]
Additional note to @aubergine10 comment, you can't change back to DRAFT when you previously marked PR as ready to merge 😉

**Comment by aubergine10** [comment]
Tested with monorail 👍  no track = lane connector not allowed = correct

image

Tested with road lane - not sure how I feel about this one. On a sharp bend, the lane connector does not allow the turn.

image
> Note: I have mod that removes lane arrows, sorry. My traffic drives on left.

I agree that there should not be a default turn there. However, I believe a user should be allowed to customise a turn there if they really want to.

It happens IRL too. For example, near where I live there is this piece of MENSA-candidate road design:

image

Traffic on B5320, in both directions, can enter Station Approach even at that crazy sharp angle. Remember, this is with vehicles driving on left (UK) so you can imagine how sharp that turn is for them. It's right next to some traffic lights, and literally on that same road is a very busy high school which gets swamped with busses and other traffic. And there's busy train station at top of that road (as name suggests) with lots of cars going up there to park. The traffic jams in mornings and evenings are truly legendary. In fact, the whole town has such nightmare roads that the only way to solve problem was to build a bypass road in the hopes of at least some traffic being able to get to work prior to the driver growing old and reaching retirement age on a single trip.

So, to summarise: For roads, default should block out sharp angled turns, but TM:PE tools (lane arrows and connectors) should allow users to create genius-tier traffic jams in their cities by going right ahead and connecting those sharp turns up and sending heavy traffic down them much to the dismay of local cims.

**Comment by aubergine10** [comment]
In fact, I have to rant some more about that junction because the whole world has to know and hopefully someone will send an international delegation of transport engineers to fix the problem and end 50 years of traffic nightmare.

Station road is a steep hill. They used to have upwards of 30 bus convoys going up there on the school run, and back then busses were always breaking down. Every few weeks the whole road system would gridlock for an entire day while a broken bus was repaired by hand with spare parts sent by train! And one time the breaks went on a bus and it plowed in to the pedestrian crossing full of kids. But that was just a minor incident...

The pedestrian crossing is at just the right place to make it almost impossible for vehicles to do far side turn in to Station Road; they can't even start turning until they are over the crossing. So end up with massive queue backed up in that direction too. And because of oncoming traffic, the pedestrian crossing spends most of the day obstructed with cars trying to turn. Each change of the traffic lights you get something like 3-5 vehicles through because it is an abomination.

So then they started trying to change school hours earlier/later, and started schemes to get flexitime in to all businesses in the two nearest CITIES. Entire work schedules of two entire CITIES were being disrupted by one freaking junction!

Traffic was trying to find all kinds of alternate routes. There used to be a tiny road that went behind the school, it was the only way to bypass that junction. So upwards of 20,000/hour cars started detouring down small residential streets, at same time busses full of school kids and residents were trying to get to/from school/work. The entire town was gridlocked for over a decade.

So, being genuis-level deisgners, the local council embarked on a traffic calming programme of truly biblical proportions. Becuase what you need in a gridlocked city is vast amounts of roadworks for 5 years, not to mention all the heavy vehicles delivering concrete and stuff.

They installed mini-roundabouts, over 2000 speed bumps (really huge ones that 2 cars can fit on lengthways), road narrowing (literally cutting down to one lane), converting 2-way streets in to a 1-way system, installing huge plant pots on sidewalks to prevent drivers using the pavements (this literally happened!), installing medians in roads that literally weren't wide enough, etc. THROUGHOUT THE ENTIRE TOWN. We have the most traffic-calmed-yet-not-remotely-calm-by-any-stretch-of-the-imagination roads on the planet.

When that didn't work, they closed the only bypass road and for best part of a year there were still tens of thousands of vehicles trying to use the back roads and then having to turn round and it was like a maze of traffic calming contraptions.

The only other route round at the time was 75 mile detour to reach nearby motorway which itself was already gridlocked and just got worse. So they started adding lanes to that, which just made the traffic jams even bigger because obviously nobody has ever invented conflict point diagrams and the motorway has like 5 junctions in the space of a few miles, all of them gridlocked.

Then satnav was invented and drivers found another route through some country roads in nearby farmland, and part of that route was through a pig farm. A PIG FARM! Can you imagine 8,000 vehicles driving through a pig farm? That literally happened for almost 5 years. My first time using satnav sent me through that pigfarm and they've literally got an A-road with street lights and signposts and stuff IN THE PIG FARM! With pigs and llamas wandering in to the road. LLAMAS AND PIGS IN THE SAME FARM FFS!

So then they decided to build a bypass along the nearby canal. But the traffic jam was so huge they had to bypass the two neighbouring towns too. IN BOTH DIRECTIONS.

They made it a 2 lane road and it just couldn't handle the traffic. So then they decided to add more lanes, but that meant filling in half the canal to make room (as there is huge business parks and, back then, steel works, preventing extending to the other side) which also required 3 rail bridges over the canal to be rebuilt, but they only rebuilt one of them and closed 3 local stations = even more road traffic.

So then they realised that ships could no longer get down the canal, which was a problem as by that time they were building a massive LOGISTICS WAREHOUSE and SHIP PORT and SPORTS STADIUM becuase they still thought more lorries and trucks and cars would really help the traffic situation. So they decided to widen the canal on the other side, which involved demolishing an entire village and moving half of an entire chemical processing plant further down the canal. During the relocation the chemical plant exploded and decimated another small village.

**All this because nobody invented real-life Move It! mod to tweak that junction or TM:PE lane connectors.**

P.S. Please send road engineers before that junction triggers WW3.

**Comment by aubergine10** [comment]
merge conflicts

**Comment by aubergine10** [comment]
Note to self: Test to see if side-effect of these changes resolves #293

**Comment by kianzarrin** [comment]
@aubergine10 I noticed you set the release version to 11.1 without this change in the trunk.
this is going to cause problems with CSURToolbox

**Comment by aubergine10** [comment]
@kianzarrin The CSUR was updated to look for `11.1.1` and released to workshop prior to updating LABS to `11.1.0` :)

**Comment by kianzarrin** [comment]
u turn issue #293 is fixed
Screenshot (464)
Screenshot (463)

**Comment by kianzarrin** [comment]
version is updated to 11.1.1. should we merge this?

**Comment by aubergine10** [comment]
I'll do some testing tonight after testing #664 ...

**Comment by aubergine10** [comment]
Tested on some vanilla monorail roads and it works well:

* Can still connect road lanes 👍 
* Non-connected monorails can't be connected 👍 

Also tested with some vanilla tram roads but I assume they have crazy big turning angle?
image

Din't try the lane connectors as there's no way to select just tram track (#354).

I've got a few more things to test, will report back in a bit.

**Comment by aubergine10** [comment]
@kvakvs @krzychu124 This PR has been playtested and ready for code review.

**Comment by aubergine10** [comment]
Edited description to provide summary of issue.
